enterprise/cluster/src/main/java/org/neo4j/cluster/logging/AsyncLogging.java

```diff
@@ -0,0 +1,146 @@
+import org.neo4j.kernel.lifecycle.LifecycleAdapter;
+import org.neo4j.logging.Log;
+import org.neo4j.logging.LogProvider;
+import org.neo4j.logging.NullLogProvider;
+import org.neo4j.logging.async.AsyncLogEvent;
+import org.neo4j.logging.async.AsyncLogProvider;
+
+import static org.neo4j.logging.async.AsyncLogEvent.process;
+import static org.neo4j.unsafe.impl.internal.dragons.FeatureToggles.flag;
+
+public class AsyncLogging extends LifecycleAdapter
+        implements Consumer<AsyncLogEvent>, ThreadFactory, AsyncEvents.Monitor
+{
+    private static final boolean ENABLED = flag( "ENABLED", true );
+
+    public static LogProvider provider( LifeSupport life, LogProvider provider )
+    {
+        if ( ENABLED )
+        {
+            if ( provider instanceof NullLogProvider )
+            {
+                return provider;
+            }
+            return new AsyncLogProvider( life.add(
+                    new AsyncLogging( provider.getLog( AsyncLogging.class ) ) ).eventSender(), provider );
+        }
+        else
+        {
+            return provider;
+        }
+    }
+
+    private final Log metaLog;
+    private final AsyncEvents<AsyncLogEvent> events;
+    private long highCount;
+    private ExecutorService executor;
+
+    AsyncLogging( Log metaLog )
+    {
+        this.metaLog = metaLog;
+        this.events = new AsyncEvents<>( this, this );
+    }
+
+    @Override
+    public void accept( AsyncLogEvent event )
+    {
+        process( event );
+    }
+
+    @Override
+    public void start()
+    {
+        highCount = 0;
+        executor = Executors.newSingleThreadExecutor( this );
+        executor.submit( events );
+    }
+
+    @Override
+    public void stop() throws InterruptedException
+    {
+        events.shutdown();
+        executor.shutdown();
+        events.awaitTermination();
+    }
+
+    @Override
+    public Thread newThread( final Runnable target )
+    {
+        return new Thread( new RerunOnFailure( target ), getClass().getSimpleName() );
+    }
+
+    @Override
+    public void eventCount( long count )
+    {
+        if ( metaLog.isDebugEnabled() )
+        {
+            if ( count > highCount )
+            {
+                metaLog.debug( "High mark increasing from %d to %d events", highCount, count );
+                highCount = count;
+            }
+        }
+    }
+
+    AsyncEventSender<AsyncLogEvent> eventSender()
+    {
+        return events;
+    }
+
+    private static class RerunOnFailure implements Runnable
+    {
+        private final Runnable target;
+
+        RerunOnFailure( Runnable target )
+        {
+            this.target = target;
+        }
+
+        @Override
+        public void run()
+        {
+            for (; ; )
+            {
+                try
+                {
+                    target.run();
+                }
+                catch ( Throwable e )
+                {
+                    e.printStackTrace();
```

Comment:
Will propose to memorize exception and in case if it the same as before do not print stack trace for it. 
To be sure that we will notice all the exceptions that we can consume we can introduce counter of exceptions, but do not really see a use case where it can be useful for now.